mlflow/server/handlers.py

```diff
@@ -2249,6 +2263,180 @@ def _abort_multipart_upload_artifact(artifact_path):
+            "tags": [_assert_map_key_present],
+        },
+    )
+    request_metadata = {e.key: e.value for e in request_message.request_metadata}
+    tags = {e.key: e.value for e in request_message.tags}
+
+    trace_info = _get_tracking_store().end_trace(
+        request_id=request_id,
+        timestamp_ms=request_message.timestamp_ms,
+        status=TraceStatus.from_proto(request_message.status),
+        request_metadata=request_metadata,
+        tags=tags,
+    )
+    response_message = EndTrace.Response(trace_info=trace_info.to_proto())
+    return _wrap_response(response_message)
+
+
+@catch_mlflow_exception
+@_disable_if_artifacts_only
+def _get_trace_info(request_id):
+    """
+    A request handler for `GET /mlflow/traces/{request_id}/info` to retrieve
+    an existing TraceInfo record from tracking store.
+    """
+    trace_info = _get_tracking_store().get_trace_info(request_id)
+    response_message = GetTraceInfo.Response(trace_info=trace_info.to_proto())
+    return _wrap_response(response_message)
+
+
+@catch_mlflow_exception
+@_disable_if_artifacts_only
+def _search_traces():
+    """
+    A request handler for `GET /mlflow/traces` to search for TraceInfo records in tracking store.
+    """
+    request_message = _get_request_message(
+        SearchTraces(),
+        schema={
+            "experiment_ids": [_assert_array, _assert_item_type_string, _assert_required],
+            "filter": [_assert_string],
+            "max_results": [_assert_intlike, lambda x: _assert_less_than_or_equal(int(x), 500)],
+            "order_by": [_assert_array, _assert_item_type_string],
+            "page_token": [_assert_string],
+        },
+    )
+    traces, token = _get_tracking_store().search_traces(
+        experiment_ids=request_message.experiment_ids,
+        filter_string=request_message.filter,
+        max_results=request_message.max_results,
+        order_by=request_message.order_by,
+        page_token=request_message.page_token,
+    )
+    response_message = SearchTraces.Response()
+    response_message.traces.extend([e.to_proto() for e in traces])
+    if token:
+        response_message.next_page_token = token
+    return _wrap_response(response_message)
+
+
+@catch_mlflow_exception
+@_disable_if_artifacts_only
+def _delete_traces():
+    """
+    A request handler for `POST /mlflow/traces/delete-traces` to delete TraceInfo records
+    from tracking store.
+    """
+    request_message = _get_request_message(
+        DeleteTraces(),
+        schema={
+            "experiment_id": [_assert_string, _assert_required],
+            "max_timestamp_millis": [_assert_intlike],
+            "max_traces": [_assert_intlike],
+            "request_ids": [_assert_array, _assert_item_type_string],
+        },
+    )
+
+    # NB: Interestingly, the field accessor for the message object returns the default
+    #   value for optional field if it's not set. For example, `request_message.max_traces`
```

Comment:
This is a weird proto design: https://protobuf.dev/programming-guides/proto2/#optional